derive-topn-from-window.md

```diff
@@ -0,0 +1,201 @@
+---
+title: 从窗口函数中推导 TopN 或 Limit
+---
+
+# 从窗口函数中推导 TopN 或 Limit
+
+[窗口函数](/functions-and-operators/window-functions.md)是一种常见的 SQL 函数。对于 ROW_NUMBER() 或者 RANK() 等编号相关的窗口函数，一种常见的用法是在进行窗口函数求值之后，对求值的结果进行过滤，例如：
+
+```sql
+SELECT * FROM (SELECT ROW_NUMBER() OVER (ORDER BY a) AS rownumber FROM t) dt WHERE rownumber <= 3
+```
+
+按照正常的 SQL 执行流程，TiDB 需要先对 `t` 表的所有数据进行排序，然后为每一行都求得相应的 `ROW_NUMBER()` 结果，最后再进行 `rownumber <= 3` 的过滤。 
+
+从 v7.0.0 开始，TiDB 支持从窗口函数中推导 TopN 或 Limit。通过该优化规则，TiDB 可以将原始 SQL 等价改写成以下形式：
+
+```sql
+WITH t_topN AS (SELECT a FROM t1 ORDER BY a LIMIT 3) SELECT * FROM (SELECT ROW_NUMBER() OVER (ORDER BY a) AS rownumber FROM t_topN) dt WHERE rownumber <= 3
+```
+
+可以看出，该优化从窗口函数与后续的过滤条件中推导出了一个 TopN 算子，相比于原始 SQL 中的 Sort 算子，TopN 算子的运行效率远高于 Sort 算子，而且 TiKV/TiFlash 支持 TopN 算子的下推，这能进一步加速改写之后的 SQL 的性能。
+
+有两种方法关闭此优化
+
+* 设置 session 变量 [tidb_opt_derive_topn](/system-variables.md#tidb_opt_derive_topn-从-v700-版本开始引入) 为 `false`
+* 可参照[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中的关闭方法。
+
+## 示例
+
+以下通过一些例子对该优化规则进行说明。
```

Comment:
consider add a plan before this optimazition?

Reply:
Already add some plans that can not apply this rule in the following examples.